File: client/archive/android/app/src/main/java/com/solutioncrafts/BriskVTU/feature/vtu/VTUApi.kt
package com.solutioncrafts.BriskVTU.feature.vtu

import kotlinx.serialization.Serializable
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query

@Serializable
data class VariationResponse(
    val variation_code: String,
    val name: String,
    val variation_amount: String,
    val fixedPrice: String
)

@Serializable
data class VariationsApiResponse(
    val status: String,
    val data: List<VariationResponse>
)

@Serializable
data class VerifyMerchantRequest(
    val billerCode: String,
    val providerServiceId: String,
    val serviceType: String
)

@Serializable
data class VerifyMerchantApiResponse(
    val status: String,
    val data: VerifyMerchantContent
)

@Serializable
data class VerifyMerchantContent(
    val Customer_Name: String? = null,
    val Status: String? = null,
    val error: String? = null
)

@Serializable
data class InitiateTransactionRequest(
    val serviceType: String,
    val serviceId: String,
    val providerServiceId: String,
    val amount: Double,
    val variationCode: String? = null
)

@Serializable
data class TransactionResponse(
    val id: String,
    val category: String,
    val serviceType: String,
    val amount: Double,
    val status: String,
    val referenceId: String,
    val createdAt: String
)

@Serializable
data class InitiateTransactionApiResponse(
    val status: String,
    val message: String,
    val data: TransactionResponse
)


interface VTUApi {
    @GET("vtpass/variations/{serviceId}")
    suspend fun getVariations(
        @Path("serviceId") serviceId: String,
        @Query("forceRefresh") forceRefresh: Boolean = false
    ): VariationsApiResponse

    @POST("vtpass/verify")
    suspend fun verifyMerchant(@Body request: VerifyMerchantRequest): VerifyMerchantApiResponse

    @POST("transactions/initiate")
    suspend fun initiateTransaction(@Body request: InitiateTransactionRequest): InitiateTransactionApiResponse
}

File: client/archive/ios/BriskVTU/Features/VTU/VTUAPI.swift
import Foundation

struct VariationResponse: Codable, Identifiable {
    var id: String { variation_code }
    let variation_code: String
    let name: String
    let variation_amount: String
    let fixedPrice: String
}

struct VerifyMerchantRequest: Codable {
    let billerCode: String
    let providerServiceId: String
    let serviceType: String
}

struct VerifyMerchantContent: Codable {
    let Customer_Name: String?
    let Status: String?
    let error: String?
}

struct InitiateTransactionRequest: Codable {
    let serviceType: String
    let serviceId: String
    let providerServiceId: String
    let amount: Double
    let variationCode: String?
}

struct TransactionResponse: Codable, Identifiable {
    let id: String
    let category: String
    let serviceType: String
    let amount: Double
    let status: String
    let referenceId: String
    let createdAt: String
}

final class VTUAPI {
    static let shared = VTUAPI()
    private let client = APIClient.shared
    
    func getVariations(serviceId: String) async throws -> [VariationResponse] {
        let res: BaseResponse<[VariationResponse]> = try await client.request(endpoint: "/vtpass/variations/\(serviceId)", method: "GET")
        return res.data ?? []
    }
    
    func verifyMerchant(request: VerifyMerchantRequest) async throws -> VerifyMerchantContent {
        let data = try JSONEncoder().encode(request)
        let res: BaseResponse<VerifyMerchantContent> = try await client.request(endpoint: "/vtpass/verify", method: "POST", body: data)
        guard let content = res.data else { throw APIError.noData }
        return content
    }
    
    func initiateTransaction(request: InitiateTransactionRequest) async throws -> TransactionResponse {
        let data = try JSONEncoder().encode(request)
        let res: BaseResponse<TransactionResponse> = try await client.request(endpoint: "/transactions/initiate", method: "POST", body: data)
        guard let transaction = res.data else { throw APIError.noData }
        return transaction
    }
}
